src/arch/metal.rs

```diff
@@ -531,6 +533,131 @@ impl Architecture for MetalArch {
             Csr::Unknown => (),
         };
     }
+
+    unsafe fn handle_virtual_load_store(instr: Instr, ctx: &mut VirtContext) {
+        // Cannot use a raw_trap_handler because of MPRV=1, addresses would be translated
+        let trap: usize = _mprv_trap_handler as usize;
+
+        // Zero out mcause to check if a trap occured during emulation
+        Self::write_csr(Csr::Mcause, 0);
+
+        let saved_satp = Self::read_csr(Csr::Satp);
+        let saved_mpp = parse_mpp_return_mode(Self::read_csr(Csr::Mstatus));
+
+        Self::set_mpp(parse_mpp_return_mode(ctx.csr.mstatus));
+        Self::write_csr(Csr::Satp, ctx.csr.satp);
+        Self::sfence_vma(None, None);
+
+        let mut _rd_value: usize; //without "_" compiler warns that the value is never read in case of a store macro
+        let addr: usize;
+        let mut cause: usize;
+        let mut mtval: usize;
+        let mut mstatus: usize;
+        let mut mip: usize;
+        let fw_pc = ctx.trap_info.mepc;
+
+        macro_rules! construct_asm {
+            ($instr:literal) => {
+                asm!(
+                    "csrr {1}, mtvec",
+                    "csrw mtvec, {mtvec}",
+
+                    "li {0}, 1",           // Set MPRV bit to 1
+                    "slli {0}, {0}, 17",
+                    "csrs mstatus, {0}",
+
+                    concat!($instr, " {rd}, 0({addr})"),
+
+                    "csrc mstatus, {0}",
+                    "csrw mtvec, {1}",
+                    out(reg) _,
+                    out(reg) _,
+                    out("t1") cause,
+                    out("t2") mtval,
+                    out("t4") mstatus,
+                    out("t6") mip,
```

Comment:
I mean we could read it, but it's going to be an address of Miralis - what would it be used for?

Reply:
You should just tell the compiler that the register might get corrupted by the assembly snippet, otherwise you might get undefined behavior.

Adding something like `out("t5") _` will fix the issue